vsp.go

```diff
@@ -57,65 +61,315 @@ func (mw *MultiWallet) NewVSPClient(vspHost string, walletID int, purchaseAccoun
+	vspInfo, err := vsp.GetInfo(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	v.ctx = ctx
-	v.vspClient.pub = vspInfo.PubKey
-	return v, nil
+	vsp.client.pub = vspInfo.PubKey
+	return vsp, nil
 }
 
-func (v *VSP) PoolFee(ctx context.Context) (float64, error) {
+//  PoolFee returns the vsp fee percentage.
+func (vsp *VSP) PoolFee(ctx context.Context) (float64, error) {
 	var vspInfo VspInfoResponse
-	err := v.vspClient.get(ctx, apiVSPInfo, &vspInfo)
+	err := vsp.client.get(ctx, apiVSPInfo, &vspInfo)
 	if err != nil {
 		return -1, err
 	}
 	return vspInfo.FeePercentage, nil
 }
 
 // PurchaseStakeTickets purchases the number of tickets passed as an argument and pays the fees for the tickets
-func (v *VSP) PurchaseTickets(ticketCount, expiryBlocks int32, passphrase []byte) error {
-	err := v.w.UnlockWallet(passphrase)
+func (vsp *VSP) PurchaseTickets(ticketCount, expiryBlocks int32, passphrase []byte) error {
+	err := vsp.w.UnlockWallet(passphrase)
 	if err != nil {
 		return translateError(err)
 	}
-	defer v.w.LockWallet()
+	defer vsp.w.LockWallet()
 
 	request := &w.PurchaseTicketsRequest{
 		Count:                int(ticketCount),
-		SourceAccount:        v.purchaseAccount,
-		Expiry:               v.w.GetBestBlock() + expiryBlocks,
+		SourceAccount:        vsp.purchaseAccount,
+		Expiry:               vsp.w.GetBestBlock() + expiryBlocks,
 		MinConf:              DefaultRequiredConfirmations,
-		VSPFeeProcess:        v.PoolFee,
-		VSPFeePaymentProcess: v.ProcessFee,
+		VSPFeeProcess:        vsp.PoolFee,
+		VSPFeePaymentProcess: vsp.ProcessFee,
+	}
+
+	networkBackend, err := vsp.w.Internal().NetworkBackend()
+	if err != nil {
+		return err
 	}
 
-	networkBackend, err := v.w.Internal().NetworkBackend()
+	_, err = vsp.w.Internal().PurchaseTickets(vsp.ctx, networkBackend, request)
 	if err != nil {
 		return err
 	}
 
-	_, err = v.w.Internal().PurchaseTickets(v.ctx, networkBackend, request)
+	return nil
+}
+
+// StartTicketBuyer starts the automatic ticket buyer.
+func (vsp *VSP) StartTicketBuyer(balanceToMaintain int64, passphrase []byte) error {
+	if balanceToMaintain < 0 {
+		return errors.New("Negative balance to maintain given")
+	}
+
+	if vsp.w.cancelAutoTicketBuyer != nil {
+		return errors.New("Ticket buyer already running")
+	}
+
+	go func() {
+
+		log.Infof("Running ticket buyer on Wallet: %s", vsp.w.Name)
+
+		ctx, cancel := vsp.w.shutdownContextWithCancel()
+		vsp.w.cancelAutoTicketBuyer = cancel
+
+		err := vsp.runTicketBuyer(ctx, balanceToMaintain, passphrase)
+		if err != nil {
+			vsp.w.cancelAutoTicketBuyer = nil
+			if ctx.Err() != nil {
+				log.Errorf("TicketBuyerV2 instance canceled, account number: %s", vsp.w.Name)
+			}
+			log.Errorf("Ticket buyer instance errored: %v", err)
+		}
+	}()
+
+	return nil
+}
+
+func (vsp *VSP) runTicketBuyer(ctx context.Context, balanceToMaintain int64, passphrase []byte) error {
+	if len(passphrase) > 0 && vsp.w.IsLocked() {
+		err := vsp.w.UnlockWallet(passphrase)
```

Comment:
This method zeroes the passphrase, making it unusable below. Please update the `UnlockWallet` method to NOT alter the passphrase, that method has no way of knowing if it will be used subsequently. Callers ideally should be the ones to zero passphrases, when they're completely done with it.